File: app/tests/test_market_clock.py
"""Tests for the IST market-session clock."""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

import pytest

from app.config import get_settings
from app.risk import market_clock as mc

IST = mc.IST


@pytest.fixture(autouse=True)
def _enforce_market_hours(monkeypatch):
    # conftest disables market-hours enforcement globally so other suites
    # aren't time-of-day dependent; these tests need it ON to exercise the
    # real entry-window gating.
    monkeypatch.setenv("ENFORCE_MARKET_HOURS", "1")
    get_settings.cache_clear()
    yield
    get_settings.cache_clear()


def _utc_for_ist(year, month, day, hh, mm) -> datetime:
    """Build a UTC-aware datetime for a given IST wall-clock time."""
    ist_dt = datetime(year, month, day, hh, mm, tzinfo=IST)
    return ist_dt.astimezone(timezone.utc)


# 2026-06-19 is a Friday (a trading weekday); 2026-06-20 is a Saturday.
FRI = (2026, 6, 19)
SAT = (2026, 6, 20)


@pytest.mark.parametrize(
    "hh,mm,open_,entry,squareoff",
    [
        (10, 0, True, True, False),    # mid-session: everything go
        (9, 20, True, False, False),   # first 15 min after open: no entry
        (15, 5, True, False, False),   # last 30 min before close: no entry
        (15, 15, True, False, True),   # after 15:10: square-off due
        (16, 0, False, False, False),  # after close
        (8, 0, False, False, False),   # before open
    ],
)
def test_session_windows(hh, mm, open_, entry, squareoff):
    now = _utc_for_ist(*FRI, hh, mm)
    assert mc.is_market_open(now) is open_
    assert mc.is_entry_window(now) is entry
    assert mc.square_off_due(now) is squareoff


def test_weekend_is_always_closed():
    now = _utc_for_ist(*SAT, 10, 0)  # Saturday mid-session-time
    assert mc.is_market_open(now) is False
    assert mc.is_entry_window(now) is False
    assert mc.square_off_due(now) is False


def test_entry_block_reason():
    assert mc.entry_block_reason(_utc_for_ist(*FRI, 10, 0)) is None
    assert mc.entry_block_reason(_utc_for_ist(*FRI, 9, 20)) == "outside_entry_window"
    assert mc.entry_block_reason(_utc_for_ist(*FRI, 15, 15)) == "post_squareoff"
    assert mc.entry_block_reason(_utc_for_ist(*FRI, 16, 0)) == "market_closed"


def test_enforce_market_hours_toggle(monkeypatch):
    # With enforcement off, the entry gate is open even at 3am IST on a
    # Saturday — but square-off still runs on the real clock.
    monkeypatch.setenv("ENFORCE_MARKET_HOURS", "0")
    get_settings.cache_clear()
    try:
        off_hours = _utc_for_ist(*SAT, 3, 0)
        assert mc.is_entry_window(off_hours) is True
        assert mc.is_market_open(off_hours) is False
    finally:
        monkeypatch.undo()
        get_settings.cache_clear()


def test_to_ist_treats_naive_as_utc():
    naive_utc = datetime(2026, 6, 19, 4, 30)  # 04:30 UTC == 10:00 IST
    ist = mc.to_ist(naive_utc)
    assert (ist.hour, ist.minute) == (10, 0)

File: cpp/tests/test_market_clock.cpp
// IST is UTC+5:30 with no DST, so every expected value here is arithmetic on
// a fixed offset -- no tz database involved, deliberately (§9 PHASE 5).
#include <cassert>
#include <cstdio>

#include "tb/config.hpp"
#include "tb/market_clock.hpp"

using namespace tb;
using namespace std::chrono;

// 2026-08-19 is a Wednesday; 2026-08-22 a Saturday.
static SysTime utc(int y, int m, int d, int hh, int mm) {
  return sys_days{year{y} / month{static_cast<unsigned>(m)} / day{static_cast<unsigned>(d)}} +
         hours{hh} + minutes{mm};
}
// IST wall-clock time on a trading Wednesday.
static SysTime ist_wed(int hh, int mm) {
  return utc(2026, 8, 19, hh, mm) - kIstOffset;
}
static SysTime ist_sat(int hh, int mm) {
  return utc(2026, 8, 22, hh, mm) - kIstOffset;
}

int main() {
  mutable_settings() = Settings{};  // defaults: 09:15-15:30, entry 09:30-15:00

  assert(to_ist(ist_wed(9, 30)).minutes == 9 * 60 + 30);
  assert(to_ist(ist_wed(9, 30)).trading_day);
  assert(!to_ist(ist_sat(9, 30)).trading_day);

  // -- market open: inclusive on both edges ---------------------------------
  assert(!is_market_open(ist_wed(9, 14)));
  assert(is_market_open(ist_wed(9, 15)));
  assert(is_market_open(ist_wed(15, 30)));
  assert(!is_market_open(ist_wed(15, 31)));
  assert(!is_market_open(ist_sat(11, 0)));  // weekend

  // -- entry window: session minus the noisy first 15 / last 30 min ---------
  assert(!is_entry_window(ist_wed(9, 29)));
  assert(is_entry_window(ist_wed(9, 30)));
  assert(is_entry_window(ist_wed(15, 0)));
  assert(!is_entry_window(ist_wed(15, 1)));
  assert(!is_entry_window(ist_sat(11, 0)));

  // -- square-off: NOT disabled by ENFORCE_MARKET_HOURS (invariant I3) ------
  assert(!square_off_due(ist_wed(15, 9)));
  assert(square_off_due(ist_wed(15, 10)));
  assert(square_off_due(ist_wed(15, 30)));
  assert(!square_off_due(ist_wed(15, 31)));

  // -- block reasons are part of the parity contract (§10.3) ----------------
  assert(!entry_block_reason(ist_wed(10, 0)).has_value());
  assert(*entry_block_reason(ist_wed(8, 0)) == "market_closed");
  assert(*entry_block_reason(ist_sat(11, 0)) == "market_closed");
  assert(*entry_block_reason(ist_wed(9, 20)) == "outside_entry_window");
  assert(*entry_block_reason(ist_wed(15, 20)) == "post_squareoff");

  // -- the gate switches off for paper/backtest, the square-off does not ----
  mutable_settings().ENFORCE_MARKET_HOURS = false;
  assert(is_entry_window(ist_sat(3, 0)));
  assert(!entry_block_reason(ist_sat(3, 0)).has_value());
  assert(!square_off_due(ist_sat(15, 20)));      // still a weekend
  assert(square_off_due(ist_wed(15, 20)));       // still fires on a trading day
  mutable_settings() = Settings{};

  std::puts("test_market_clock OK");
  return 0;
}
